android/app/src/main/java/com/msupplymobile/Ble/BlueMaestroDevice.java

```diff
@@ -0,0 +1,200 @@
+            if (newState == BluetoothProfile.STATE_CONNECTED) {
+                gatt.discoverServices();
+                connected = true;
+            } else if (!connected && retryCount < NUMBER_OF_RETRIES) {
+                retryCount++;
+                gatt.close();
+                connectGattWithTimeout();
+            } else {
+                gatt.close();
+                completeCommand();
+            }
+        }
+        /**
+         * After connecting and discovering services, using the TX characteristic and descriptor,
+         * Write to the descriptor for initiating a UART connection.
+         */
+        @Override
+        public void onServicesDiscovered(BluetoothGatt gatt, int status) {
+            try {
+                BluetoothGattCharacteristic inCharasteristic = gatt.getService(GATT_SERVICE)
+                                                                   .getCharacteristic(IN_CHARASTERISTIC);
+                gatt.setCharacteristicNotification(inCharasteristic, true);
+                BluetoothGattDescriptor inDescriptor = inCharasteristic.getDescriptor(IN_DESCRIPTOR);
+                inDescriptor.setValue(BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE);
+                gatt.writeDescriptor((inDescriptor));
+            } catch (Exception e) {
+                notifyListener(new MsupplyException(ErrorCode.E_IN_DESC_WRITE));
+            }
+        }
+        // Write to the RX characteristic, to receive data from the onCharacteristicChanged callback.
+        @Override
+        public void onDescriptorWrite(BluetoothGatt gatt, BluetoothGattDescriptor descriptor, int status) {
+            if (descriptor.getUuid().equals(IN_DESCRIPTOR)) {
+                BluetoothGattCharacteristic outCharacteristic = gatt.getService(GATT_SERVICE)
+                                                                    .getCharacteristic(OUT_CHARASTERISTIC);
+                outCharacteristic.setValue(command);
+                boolean writeResult = gatt.writeCharacteristic(outCharacteristic);
+                if (!writeResult) notifyListener(new MsupplyException(ErrorCode.E_OUT_CHAR_WRITE));
+            }
+        }
+
+        /**
+         * On writing to the RX characteristic, continually receive a stream of data from the device
+         * until the connection closes. Store a flattened array of bytes for this stream.
+         */
+        @Override
+        public void onCharacteristicChanged(BluetoothGatt gatt, BluetoothGattCharacteristic characteristic) {
+            byte response[] = characteristic.getValue();
+            ByteArrayOutputStream output = new ByteArrayOutputStream();
+            try{
+                output.write(commandResult);
+                output.write(response);
+            } catch(Exception exception){
+                notifyListener(new MsupplyException(ErrorCode.E_DOWNLOADING_LOGS, exception));
+                
+            }
+            commandResult = output.toByteArray();
```

Comment:
i see you are flattening a 2 d array response here, I don't really see anything wrong with that apart from, is there a chance that LOGS_HEADER_OFFSET is not always 15 ? Is this how the blue maestro pps do it ?

Reply:
It's not how they do it, but I tried to do pretty extensive testing. I think if we can't rely on *some* form of consistency it's pretty hard to decode anything